event-svc/src/event/validator/grouped.rs

```diff
@@ -0,0 +1,125 @@
+use ceramic_event::unvalidated::{self, signed};
+use ipld_core::ipld::Ipld;
+
+use super::UnvalidatedEvent;
+
+#[derive(Debug)]
+pub enum ValidationNeeded {
+    SignedInit(SignedInit),
+    SignedData(SignedData),
+    Time(Time),
+    None(Unsigned),
+}
+
+impl From<UnvalidatedEvent> for ValidationNeeded {
+    fn from(value: UnvalidatedEvent) -> Self {
+        match value.event.as_ref() {
+            unvalidated::Event::Time(_) => Self::Time(Time(value)),
+            unvalidated::Event::Signed(signed) => match signed.payload() {
+                unvalidated::Payload::Data(_) => Self::SignedData(SignedData(value)),
+                unvalidated::Payload::Init(_) => Self::SignedInit(SignedInit(value)),
+            },
+            unvalidated::Event::Unsigned(_) => Self::None(Unsigned(value)),
+        }
+    }
+}
+
+macro_rules! unvalidated_wrapper {
+    ($name: ident) => {
+        #[derive(Debug)]
+        pub(crate) struct $name(UnvalidatedEvent);
+
+        impl $name {
+            #[allow(dead_code)]
+            /// Get the inner tuple struct value
+            pub fn into_inner(self) -> UnvalidatedEvent {
+                self.0
+            }
+
+            #[allow(dead_code)]
+            /// Get the inner tuple struct value by reference
+            pub fn as_inner(&self) -> &UnvalidatedEvent {
+                &self.0
+            }
+        }
+    };
+}
+
+macro_rules! signed_event {
+    ($name: ident) => {
+        impl $name {
+            pub fn as_signed(&self) -> &signed::Event<Ipld> {
+                match self.0.event.as_ref() {
+                    unvalidated::Event::Time(_) => unreachable!("time event is not signed"),
+                    unvalidated::Event::Signed(s) => s,
+                    unvalidated::Event::Unsigned(_) => unreachable!("unsigned event is not signed"),
+                }
+            }
+        }
+    };
+}
+
+impl Time {
+    #[allow(dead_code)]
+    pub fn as_time(&self) -> &unvalidated::TimeEvent {
+        match self.0.event.as_ref() {
+            unvalidated::Event::Time(t) => t,
+            unvalidated::Event::Signed(_) => unreachable!("signed event event is not time"),
+            unvalidated::Event::Unsigned(_) => unreachable!("unsigned event is not time"),
+        }
+    }
+}
+
+unvalidated_wrapper!(SignedData);
+unvalidated_wrapper!(SignedInit);
+unvalidated_wrapper!(Time);
+unvalidated_wrapper!(Unsigned);
+
+signed_event!(SignedData);
+signed_event!(SignedInit);
+
+#[derive(Debug)]
+pub struct GroupedEvents {
+    pub time_events: Vec<Time>,
+    pub signed_events: SignedEvents,
+}
+
+#[derive(Debug)]
+pub struct SignedEvents {
+    pub data: Vec<SignedData>,
+    pub init: Vec<SignedInit>,
+    pub unsigned: Vec<Unsigned>,
```

Comment:
yeah, we could but we'd need to pass it to the signed event validator still. maybe the name isn't great, but it's intended to partition everything into the two validation buckets, and event signatures need all the init events (signed and unsigned) since they aren't in the database yet to look up

Reply:
why do unsigned events need to go to the SignedEventValidator?  Can't they just skip validation entirely?

Reply:
oh I see, they need to be passed in not to be validated, but as context for validating other events in the batch.  Okay makes sense, probably just a comment to that effect will help